diff --git a/package/debian/nosh-run-system-manager.post_install.extra b/package/debian/nosh-run-system-manager.post_install.extra
index 25c03dd..c3a9b47 100644
--- a/package/debian/nosh-run-system-manager.post_install.extra
+++ b/package/debian/nosh-run-system-manager.post_install.extra
@@ -1,12 +1,9 @@
 # Extra post-install for nosh-run-system-manager
 # vim: set filetype=sh:
 
-if test -z "$2"		# No prior version installed/left configured.
-then
 	for i in - -usr -var -home 
 	do
 		/bin/rm -f -r /etc/service-bundles/services/fsck@$i/
 		/bin/rm -f -r /etc/service-bundles/services/mount@$i/
 	done
-fi
 redo -C /etc/system-control/convert all
diff --git a/package/debian/nosh-run-system-manager.post_upgrade.extra b/package/debian/nosh-run-system-manager.post_upgrade.extra
index 1fce0b2..4c1e018 100644
--- a/package/debian/nosh-run-system-manager.post_upgrade.extra
+++ b/package/debian/nosh-run-system-manager.post_upgrade.extra
@@ -1,12 +1,9 @@
 # Extra post-upgrade for nosh-run-system-manager
 # vim: set filetype=sh:
 
-if test -z "$2"		# No prior version installed/left configured.
-then
 	for i in - -usr -var -home 
 	do
 		/bin/rm -f -r /etc/service-bundles/services/fsck@$i/
 		/bin/rm -f -r /etc/service-bundles/services/mount@$i/
 	done
-fi
 redo -C /etc/system-control/convert all
diff --git a/package/debian/nosh-service-management.post_remove.extra b/package/debian/nosh-service-management.post_remove.extra
index 4086aa9..79e0316 100644
--- a/package/debian/nosh-service-management.post_remove.extra
+++ b/package/debian/nosh-service-management.post_remove.extra
@@ -1,4 +1,4 @@
 # Extra post-remove for nosh-service-management
 # vim: set filetype=sh:
 
-dpkg-divert --package nosh-service-management --remove --rename /usr/lib/dbus-1.0/dbus-daemon-launch-helper
+mv -v /usr/lib/dbus-1.0/dbus-daemon-launch-helper.old /usr/lib/dbus-1.0/dbus-daemon-launch-helper
diff --git a/package/debian/nosh-service-management.pre_install.extra b/package/debian/nosh-service-management.pre_install.extra
index f4657c0..c35b929 100644
--- a/package/debian/nosh-service-management.pre_install.extra
+++ b/package/debian/nosh-service-management.pre_install.extra
@@ -1,4 +1,4 @@
 # Extra pre-install for nosh-service-management
 # vim: set filetype=sh:
 
-dpkg-divert --package nosh-service-management --add --rename /usr/lib/dbus-1.0/dbus-daemon-launch-helper
+mv -v /usr/lib/dbus-1.0/dbus-daemon-launch-helper /usr/lib/dbus-1.0/dbus-daemon-launch-helper.old
diff --git a/package/debian/nosh-service-management.pre_upgrade.extra b/package/debian/nosh-service-management.pre_upgrade.extra
index 71cb641..6d994d1 100644
--- a/package/debian/nosh-service-management.pre_upgrade.extra
+++ b/package/debian/nosh-service-management.pre_upgrade.extra
@@ -1,4 +1,4 @@
 # Extra pre-upgrade for nosh-service-management
 # vim: set filetype=sh:
 
-dpkg-divert --package nosh-service-management --add --rename /usr/lib/dbus-1.0/dbus-daemon-launch-helper
+mv -v /usr/lib/dbus-1.0/dbus-daemon-launch-helper /usr/lib/dbus-1.0/dbus-daemon-launch-helper.old
