diff --git a/package/debian/nosh-run-system-manager.post_install.extra b/package/debian/nosh-run-system-manager.post_install.extra
index cacf3c1..d30d4a1 100644
--- a/package/debian/nosh-run-system-manager.post_install.extra
+++ b/package/debian/nosh-run-system-manager.post_install.extra
@@ -1,11 +1,8 @@
 # Extra post-install for nosh-run-system-manager
 
-if test -z "$2"		# No prior version installed/left configured.
-then
 	for i in - -usr -var -home 
 	do
 		/bin/rm -f -r /etc/service-bundles/services/fsck@$i/
 		/bin/rm -f -r /etc/service-bundles/services/mount@$i/
 	done
-fi
 redo -C /etc/system-control/convert all
diff --git a/package/debian/nosh-run-system-manager.post_upgrade.extra b/package/debian/nosh-run-system-manager.post_upgrade.extra
index afc1354..7a926e1 100644
--- a/package/debian/nosh-run-system-manager.post_upgrade.extra
+++ b/package/debian/nosh-run-system-manager.post_upgrade.extra
@@ -1,11 +1,8 @@
 # Extra post-upgrade for nosh-run-system-manager
 
-if test -z "$2"		# No prior version installed/left configured.
-then
 	for i in - -usr -var -home 
 	do
 		/bin/rm -f -r /etc/service-bundles/services/fsck@$i/
 		/bin/rm -f -r /etc/service-bundles/services/mount@$i/
 	done
-fi
 redo -C /etc/system-control/convert all
diff --git a/package/debian/nosh-service-management.post_remove.extra b/package/debian/nosh-service-management.post_remove.extra
index ef85f0f..0d1c6db 100644
--- a/package/debian/nosh-service-management.post_remove.extra
+++ b/package/debian/nosh-service-management.post_remove.extra
@@ -1,3 +1,3 @@
 # Extra post-remove for nosh-service-management
 
-dpkg-divert --package nosh-service-management --remove --rename /usr/lib/dbus-1.0/dbus-daemon-launch-helper
+mv -v /usr/lib/dbus-1.0/dbus-daemon-launch-helper.old /usr/lib/dbus-1.0/dbus-daemon-launch-helper
diff --git a/package/debian/nosh-service-management.pre_install.extra b/package/debian/nosh-service-management.pre_install.extra
index 546254e..45f0576 100644
--- a/package/debian/nosh-service-management.pre_install.extra
+++ b/package/debian/nosh-service-management.pre_install.extra
@@ -1,3 +1,3 @@
 # Extra pre-install for nosh-service-management
 
-dpkg-divert --package nosh-service-management --add --rename /usr/lib/dbus-1.0/dbus-daemon-launch-helper
+mv -v /usr/lib/dbus-1.0/dbus-daemon-launch-helper /usr/lib/dbus-1.0/dbus-daemon-launch-helper.old
diff --git a/package/debian/nosh-service-management.pre_upgrade.extra b/package/debian/nosh-service-management.pre_upgrade.extra
index ab7e91e..837eac7 100644
--- a/package/debian/nosh-service-management.pre_upgrade.extra
+++ b/package/debian/nosh-service-management.pre_upgrade.extra
@@ -1,3 +1,3 @@
 # Extra pre-upgrade for nosh-service-management
 
-dpkg-divert --package nosh-service-management --add --rename /usr/lib/dbus-1.0/dbus-daemon-launch-helper
+mv -v /usr/lib/dbus-1.0/dbus-daemon-launch-helper /usr/lib/dbus-1.0/dbus-daemon-launch-helper.old
