diff --git a/source/services/dbus-broker.service b/source/services/dbus-broker.service
index 91106ca..d882848 100644
--- a/source/services/dbus-broker.service
+++ b/source/services/dbus-broker.service
@@ -9,7 +9,7 @@ After=machine-id.service extra-shared-libraries.target
 [Service]
 RuntimeDirectory=dbus
 RuntimeDirectoryMode=0755
-ExecStartPre=setfacl -m 'u:messagebus:rw' "%t/service-manager/control"
+ExecStartPre=setfacl -m 'u:dbus:rw' "%t/service-manager/control"
 ExecStartPre=install -d -m 0755 -- /var/db/dbus
 ExecStartPre=sh -c 'test -r /var/db/dbus/machine-id || ln -s /etc/machine-id /var/db/dbus/'
 ExecStart=%p-launch -v --scope system --listen inherit
diff --git a/source/services/dbus-daemon.service.linux b/source/services/dbus-daemon.service.linux
index e1787c8..98da3e0 100644
--- a/source/services/dbus-daemon.service.linux
+++ b/source/services/dbus-daemon.service.linux
@@ -9,7 +9,7 @@ After=machine-id.service extra-shared-libraries.target
 [Service]
 RuntimeDirectory=dbus
 RuntimeDirectoryMode=0755
-ExecStartPre=setfacl -m 'u:messagebus:rw' "%t/service-manager/control"
+ExecStartPre=setfacl -m 'u:dbus:rw' "%t/service-manager/control"
 ExecStartPre=install -d -m 0755 -- /var/db/dbus
 ExecStartPre=sh -c 'test -r /var/db/dbus/machine-id || ln -s /etc/machine-id /var/db/dbus/'
 # The --address=systemd: option is the closest to our UCSPI-LOCAL.
diff --git a/source/services/system-wide.conf.linux b/source/services/system-wide.conf.linux
index 3f7e475..63d53f7 100644
--- a/source/services/system-wide.conf.linux
+++ b/source/services/system-wide.conf.linux
@@ -1,6 +1,6 @@
 <!DOCTYPE busconfig PUBLIC "-//freedesktop//DTD D-Bus Bus Configuration 1.0//EN" "http://www.freedesktop.org/standards/dbus/1.0/busconfig.dtd">
 <busconfig>
-  <include>/etc/dbus-1/system.conf</include>
+  <include>/usr/share/dbus-1/system.conf</include>
   <!-- This is a helper that is used to launch system-wide services -->
-  <servicehelper>/usr/local/bin/dbus-daemon-launch-helper</servicehelper>
+  <servicehelper>/usr/bin/dbus-daemon-launch-helper</servicehelper>
 </busconfig>
